spec/generics.rst

```diff
@@ -0,0 +1,2503 @@
+
+
+User-Defined Generic Classes
+""""""""""""""""""""""""""""
+
+Just as defining a class as inheriting from ``Generic[T]`` makes a class generic
+for a single parameter (when ``T`` is a ``TypeVar``\ ), defining a class as
+inheriting from ``Generic[P]`` makes a class generic on
+``parameters_expression``\ s (when ``P`` is a ``ParamSpec``).
+
+.. code-block::
+
+   T = TypeVar("T")
+   P_2 = ParamSpec("P_2")
+
+   class X(Generic[T, P]):
+     f: Callable[P, int]
+     x: T
+
+   def f(x: X[int, P_2]) -> str: ...                    # Accepted
+   def f(x: X[int, Concatenate[int, P_2]]) -> str: ...  # Accepted
+   def f(x: X[int, [int, bool]]) -> str: ...            # Accepted
+   def f(x: X[int, ...]) -> str: ...                    # Accepted
+   def f(x: X[int, int]) -> str: ...                    # Rejected
+
+Or, equivalently, using the built-in syntax for generics in Python 3.12
+and higher::
+
+  class X[T, **P]:
+    f: Callable[P, int]
+    x: T
+
+By the rules defined above, spelling a concrete instance of a class generic
+with respect to only a single ``ParamSpec`` would require unsightly double
+brackets.  For aesthetic purposes we allow these to be omitted.
+
+.. code-block::
+
+   class Z(Generic[P]):
+     f: Callable[P, int]
+
+   def f(x: Z[[int, str, bool]]) -> str: ...   # Accepted
+   def f(x: Z[int, str, bool]) -> str: ...     # Equivalent
+
+   # Both Z[[int, str, bool]] and Z[int, str, bool] express this:
+   class Z_instantiated:
+     f: Callable[[int, str, bool], int]
+
+Semantics
+"""""""""
+
+The inference rules for the return type of a function invocation whose signature
+contains a ``ParamSpec`` variable are analogous to those around
+evaluating ones with ``TypeVar``\ s.
+
+.. code-block::
+
+   def changes_return_type_to_str(x: Callable[P, int]) -> Callable[P, str]: ...
+
+   def returns_int(a: str, b: bool) -> int: ...
+
+   f = changes_return_type_to_str(returns_int) # f should have the type:
+                                               # (a: str, b: bool) -> str
+
+   f("A", True)               # Accepted
+   f(a="A", b=True)           # Accepted
+   f("A", "A")                # Rejected
+
+   expects_str(f("A", True))  # Accepted
+   expects_int(f("A", True))  # Rejected
+
+Just as with traditional ``TypeVars``\ , a user may include the same
+``ParamSpec`` multiple times in the arguments of the same function,
+to indicate a dependency between multiple arguments.  In these cases a type
+checker may choose to solve to a common behavioral supertype (i.e. a set of
+parameters for which all of the valid calls are valid in both of the subtypes),
+but is not obligated to do so.
+
+.. code-block::
+
+   P = ParamSpec("P")
+
+   def foo(x: Callable[P, int], y: Callable[P, int]) -> Callable[P, bool]: ...
+
+   def x_y(x: int, y: str) -> int: ...
+   def y_x(y: int, x: str) -> int: ...
+
+   foo(x_y, x_y)  # Should return (x: int, y: str) -> bool
+
+   foo(x_y, y_x)  # Could return (__a: int, __b: str) -> bool
```

Comment:
```suggestion
   foo(x_y, y_x)  # Could return (__a: int, __b: str, /) -> bool
```
Just to really drive home the fact that these are positional only arguments

Reply:
We shouldn't use the double underscore convention at all though.

Reply:
Usually I'd agree and my first instinct was to change it to `(a: int, b: str, /)`, however we aren't really defining a function here, we just want to communicate that the signature of the combined `ParamSpec` should be positional only, so using double underscores in this case I think helps communicate that the name doesn't really matter a bit better. I was also considering `(int, str) -> bool`, because I think that's how mypy would display that signature, but that only works because mypy uses the qualified names for the types to disambiguate them from parameter names.

Ideally we would have a common way to describe signatures that we can also use in `Callable` or other generics with `ParamSpec` where the positional only arguments don't have a name at all and the mixed or keyword only arguments do.

Reply:
Yeah, another thing I considered was `(_: int, _:str, /) -> bool`, although that would be invalid syntax once you would try it in a `def`.

Reply:
I'm going to change this in a separate PR.